### ISSUE TITLE
Bug: Multiple rootDirs in tsconfig.json causes "Unsafe assignment of an `any` value" for monorepo

### ISSUE THREAD

**Issue by squirmy** [eslint/eslint#17700]
### Environment

Node version: v20.8.0
npm version: 10.1.0
Local ESLint version: 8.52.0
Global ESLint version: n/a
Operating System: macOS

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

Please see: https://github.com/squirmy/eslint-root-dirs-issue
</details>

I have a monorepo using typescript and am experience issues when importing from a reusable package if it contains types which exist in another `rootDir`

I have provided a repository with a reproduction in it. A summary of the important parts are below.

`@eslint-root-dirs-issue/one/types/src/index.ts`
```ts
export interface TypeFromAnotherRoot {
  text: string
}
```

`@eslint-root-dirs-issue/one/src/index.ts`
```ts
import type { TypeFromAnotherRoot } from './types'

export interface TypeFromSourceRoot {
  text: string
}

export function good() {
  return { text: '' } satisfies TypeFromSourceRoot
}

export function bad() {
  return { text: '' } satisfies TypeFromAnotherRoot
}

export function badExplicit(): TypeFromAnotherRoot {
  return { text: '' }
}
```

`@eslint-root-dirs-issue/one/tsconfig.json`
```json
{
  "extends": [
    "@eslint-root-dirs-issue/typescript-config/tsconfig.json"
  ],
  "compilerOptions": {
    "rootDirs": [
      ".",
      "./types"
    ]
  },
  "include": [
    "./types/**/*.d.ts",
    "./eslint.config.js",
    "./src/**/*.ts"
  ],
  "exclude": [
    "./node_modules/**"
  ]
}
```

`@eslint-root-dirs-issue/two/src/index.ts`
```ts
import { bad, badExplicit, good } from '@eslint-root-dirs-issue/one'

export function good_works() {
  // eslint-disable-next-line unused-imports/no-unused-vars
  const { text } = good()
}

export function bad_fails() {
  // eslint-disable-next-line unused-imports/no-unused-vars
  const { text } = bad()
}

export function badExplicit_fails() {
  // eslint-disable-next-line unused-imports/no-unused-vars
  const { text } = badExplicit()
}
```

```ts
import { bad, badExplicit, good } from '@eslint-root-dirs-issue/one'

export function good_works() {
  // eslint-disable-next-line unused-imports/no-unused-vars
  const { text } = good()
}

export function bad_fails() {
  // eslint-disable-next-line unused-imports/no-unused-vars
  const { text } = bad()
}

export function badExplicit_fails() {
  // eslint-disable-next-line unused-imports/no-unused-vars
  const { text } = badExplicit()
}
```

### What did you expect to happen?

eslint succeeds with no errors.

### What actually happened?

```
> @eslint-root-dirs-issue/two@1.0.0 lint /Users/squirmy/code/eslint-root-dirs-issue/packages/two
> eslint .


/Users/squirmy/code/eslint-root-dirs-issue/packages/two/src/index.ts
    10:9  error  Unsafe assignment of an `any` value  ts/no-unsafe-assignment
    15:9  error  Unsafe assignment of an `any` value  ts/no-unsafe-assignment

✖ 2 problems (2 errors, 0 warnings)

 ELIFECYCLE  Command failed with exit code 1.
```

### Link to Minimal Reproducible Example

https://github.com/squirmy/eslint-root-dirs-issue

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I have a feeling this might be caused because each project is being linted individually and I'm not sure how to tell the consumer of the package that it should also include the rootDir. So there's a chance that this could be fixed by adding configuration to the consumer.

**Comment by nzakas** [comment]
This is actually an issue with [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) rather than ESLint itself. Please open an issue there.
